test/test_common/simulated_time_system.h

```diff
@@ -0,0 +1,92 @@
+#pragma once
+
+#include "envoy/event/timer.h"
+
+#include "common/common/lock_guard.h"
+#include "common/common/thread.h"
+#include "common/common/utility.h"
+
+namespace Envoy {
+namespace Event {
+
+// Represents a simulated time system, where time is advanced by calling
+// sleep(), setSystemTime(), or setMonotonicTime(). systemTime() and
+// monotonicTime() are maintained in the class, and alarms are fired in response
+// to adjsutments in time.
+class SimulatedTimeSystem : public TimeSystem {
+public:
+  SimulatedTimeSystem();
+
+  // TimeSystem
+  SchedulerPtr createScheduler(Libevent::BasePtr&) override;
+
+  // TimeSource
+  SystemTime systemTime() override;
+  MonotonicTime monotonicTime() override;
+
+  /**
+   * Advances time forward by the specified duration, running any timers
+   * along the way that have been scheduled to fire.
+   *
+   * @param duration The amount of time to sleep, expressed in any type that
+   * can be added to duration-cast to MonotonicTime::duration.
```

Comment:
nit: I had a hard time parsing this, maybe switch the hypen to "when"?

Reply:
reworded; to:
```
   * @param duration The amount of time to sleep, expressed in any type that
   * can be duration_casted to MonotonicTime::duration.
```